python/lsst/daf/butler/core/config.py

```diff
@@ -349,32 +406,102 @@ def names(self, topLevelOnly=False):
-        names : `list`
-            List of all names present in the `Config`.
-
-        Notes
-        -----
-        This is different than the built-in method `dict.keys`, which will
-        return only the first level keys.
+        names : `list` of `tuple` of `str`
+            List of all names present in the `Config` where each element
+            in the list is a `tuple` of strings representing the hierarchy.
         """
         if topLevelOnly:
-            return list(self.keys())
+            return list((k,) for k in self)
 
-        def getKeys(d, keys, base):
+        def getKeysAsTuples(d, keys, base):
             if isinstance(d, collections.Sequence):
                 theseKeys = range(len(d))
             else:
                 theseKeys = d.keys()
             for key in theseKeys:
                 val = d[key]
-                levelKey = f"{base}.{key}" if base is not None else key
+                levelKey = base + (key,) if base is not None else (key,)
                 keys.append(levelKey)
                 if isinstance(val, (collections.Mapping, collections.Sequence)) and not isinstance(val, str):
-                    getKeys(val, keys, levelKey)
+                    getKeysAsTuples(val, keys, levelKey)
         keys = []
-        getKeys(self.data, keys, None)
+        getKeysAsTuples(self.data, keys, None)
         return keys
 
+    def names(self, topLevelOnly=False, delimiter=None):
+        """Get the delimited name of all the keys in the hierarchy.
+
+        The values returned from this method are guaranteed to be usable
+        to access items in the configuration object.
+
+        Parameters
+        ----------
+        topLevelOnly : `bool`, optional
+            If False, the default, a full hierarchy of names is returned.
+            If True, only the top level are returned.
+        delimiter : `str`, optional
+            Delimiter to use when forming the keys.  The delimiter must
+            not be present in any of the keys.  If `None` given the delimiter
+            will be automatically provided.
+
+        Returns
+        -------
+        names : `list`
+            List of all names present in the `Config`.
+
+        Notes
+        -----
+        This is different than the built-in method `dict.keys`, which will
+        return only the first level keys.
+
+        Raises
+        ------
+        ValueError
+            The externally specified delimiter is unsuitable because a key
+            clashes with the value.
+        """
+        if topLevelOnly:
+            return list(self.keys())
+
+        # Get all the tuples of hierarchical keys
+        nameTuples = self.nameTuples()
+
+        # Form big string for easy check of delimiter clash
+        combined = "".join("".join(str(s) for s in k) for k in nameTuples)
+
+        if delimiter is not None:
+            if delimiter in combined:
+                # For better error message we need to report the clashing key
```

Comment:
I was hoping the backslash-quoting would allow you to use the provided delimiter even in this case, by quoting it.

Reply:
Ah. Yes. You are right. I could escape the delimiter in the keys in this case.